src/core/matchers/toHaveSize.js

```diff
@@ -0,0 +1,57 @@
+getJasmineRequireObj().toHaveSize = function(j$) {
+  /**
+   * {@link expect} the actual size to be equal to the expected, using array-like length or object keys size.
+   * @function
+   * @name matchers#toHaveSize
+   * @since 3.5.1
+   * @param {Object} expected - Expected size
+   * @example
+   * array = [1,2];
+   * expect(array).toHaveSize(2);
+   */
+  function toHaveSize(matchersUtil) {
+    return {
+      compare: function(actual, expected) {
+        var result = {
+            pass: false
+          },
+          diffBuilder = j$.DiffBuilder({prettyPrinter: matchersUtil.pp});
+
+        // Avoid misleading collections size matching
+        if (actual instanceof WeakSet
+          || actual instanceof WeakMap
+          || actual instanceof DataView)
+          throw new Error('Cannot get size of ' + actual + '.');
+
+        // Ref https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects
+        if (Array.isArray(actual))
```

Comment:
If we're going to call out this many special cases, I'd like to see specific checks for each of them in the test. If that seems excessive, I'd be happy to see a more lenient check (e.g. does the `actual` have a `length` or `size`) and then just add some tests for those features.

Reply:
Some js framework uses length property to detects arrays, but this is quite dangerous.
I could use the lodash way, which provides some additionnal checks
https://lodash.com/docs/4.17.15#isArrayLike
I agree that would simplify.

Reply:
Since we're also having this work for strings, it might actually be fine just to check for a `length` property that is a `number` and then some specific checks for `Map` and `Set`

Reply:
see abe38429f5ee4e40727b8a2c22e36ebc9bf5b10e